### PR TITLE
Introduce Boolean Coercion

### PR DESCRIPTION
**Pull request by jayzhan211**
## Which issue does this PR close?

Ref #8302

### CODE REVIEW

File: datafusion/expr/src/type_coercion/binary.rs [hunk 1]
```diff
@@ -353,6 +354,20 @@ fn string_temporal_coercion(
     }
 }
 
+/// Coerce `Boolean` to other larger types, like Numeric as `1` or String as "1"
+fn bool_coercion(lhs_type: &DataType, rhs_type: &DataType) -> Option<DataType> {
+    match (lhs_type, rhs_type) {
```
Comment (alamb) [inline datafusion/expr/src/type_coercion/binary.rs:359]: I think this also needs to check when `rhs_type` is a `DataType::Boolean` as well. 

I would expect both of the following queries to work and return the same thing

```
sDataFusion CLI v33.0.0
❯ select 1 = true;
Error during planning: Cannot infer common argument type for comparison operation Int64 = Boolean
❯ select true = 1;
+--------------------------+
| Boolean(true) = Int64(1) |
+--------------------------+
| true                     |
+--------------------------+
1 row in set. Query took 0.006 seconds.
```

File: datafusion/sqllogictest/test_files/array.slt [hunk 2]
```diff
@@ -557,10 +557,16 @@ select column1[0:5], column2[0:3], column3[0:9] from arrays;
 ## make_array (aliases: `make_list`)
 
 # make_array scalar function #1
-query ???
-select make_array(1, 2, 3), make_array(1.0, 2.0, 3.0), make_array('h', 'e', 'l', 'l', 'o');
-----
-[1, 2, 3] [1.0, 2.0, 3.0] [h, e, l, l, o]
+query ??????
```
Comment (alamb) [inline datafusion/sqllogictest/test_files/array.slt:560]: Can you also add some tests for other uses of this logic (not just in make_array) such as comparisons

I notice that postgres doesn't handle boolean coercion 

```
postgres=# select true = 1;
ERROR:  operator does not exist: boolean = integer
LINE 1: select true = 1;
                    ^
HINT:  No operator matches the given name and argument types. You might need to add explicit type casts.
```

However, after this PR datafusion does:

```sql
DataFusion CLI v33.0.0
❯ select true = 1;
+--------------------------+
| Boolean(true) = Int64(1) |
+--------------------------+
| true                     |
+--------------------------+
1 row in set. Query took 0.006 seconds.
```

**Review by alamb — COMMENTED** [review]
Thank you @jayzhan211  -- this is looking pretty close. I left some suggestions -- let me know what you think

File: datafusion/sqllogictest/test_files/array.slt (hunk 2 above)
Comment (jayzhan211) [reply datafusion/sqllogictest/test_files/array.slt:560]: Duckdb support boolean coercion, we can consider follow it. If it does not break the overall design, I think we can support it too.

File: datafusion/physical-expr/src/expressions/case.rs [hunk 3]
```diff
@@ -828,7 +828,8 @@ mod tests {
             None,
             schema.as_ref(),
         );
-        assert!(expr.is_err());
+        // expr is ok since boolean coercion is supported
```
Comment (alamb) [inline datafusion/physical-expr/src/expressions/case.rs:831]: I think we lost the negative coverage here -- can. you please change the test back to checking that this is an error?

Perhaps you could change it to 

```
CASE WHEN a = 'foo' THEN timestamp(500ns) WHEN a = 'bar' THEN true END
```

As bool --> timestamp coercion isn't supported

File: datafusion/physical-expr/src/expressions/case.rs [hunk 4]
```diff
@@ -799,69 +799,6 @@ mod tests {
         Ok(batch)
     }
 
-    #[test]
```
Comment (jayzhan211) [inline datafusion/physical-expr/src/expressions/case.rs:802]: move to slt

File: datafusion/sql/tests/sql_integration.rs [hunk 5]
```diff
@@ -2190,18 +2190,6 @@ fn union_with_aliases() {
     quick_test(sql, expected);
 }
 
-#[test]
-fn union_with_incompatible_data_types() {
```
Comment (jayzhan211) [inline datafusion/sql/tests/sql_integration.rs:2194]: move to slt

**Comment by jayzhan211** [comment]
What determine the order of union
```
[SQL] SELECT 'a' a UNION ALL SELECT true a
error: test failed, to rerun pass `-p datafusion-sqllogictest --test sqllogictests`
[Diff] (-expected|+actual)

-   true
Caused by:
-   a
  process didn't exit successfully: `/Users/runner/work/arrow-datafusion/arrow-datafusion/target/debug/deps/sqllogictests-4f4c8273301f5ec4` (exit status: 1)
+   a
+   true
at test_files/union.slt:161
```

I got true, a in local. but ci failed on since the order is not the same as mine.

**Comment by jackwener** [comment]
`Boolean Coercion` is a dangerous behavior. It will cause some strange behavior.

I suggest to we'd better to keep same behavior with `pg`. 

There is a paper about `type theory` in `type coercion`.

https://www.microsoft.com/en-us/research/wp-content/uploads/2016/02/icfp051-swamy.pdf

**Comment by jayzhan211** [comment]
> `Boolean Coercion` is a dangerous behavior. It will cause some strange behavior.
> 
> such as:
> 
> ```
> If there is some implicit type cast:
> 1. true = 'true'
> 2. true = 1
> 
> But true = 1 = '1' is conflict with true = 'true'
> ```
> 
> I suggest to we'd better to keep same behavior with `pg`.

I agree it causes some confusing behavior.
```
query ?
select make_array(true, 1, 'true');
----
[true, 1, true]

query ?
select make_array(true, 'true', 1);
----
[true, true, 1]

query ?
select make_array(true, 1);
----
[1, 1]
```

Then, why do DuckDB introduce this 🤔

**Pull request closed without merging**